src/client/handlers/AuditEventsGet.ts

```diff
@@ -42,8 +43,8 @@ class AuditEventsGet extends ServerHandler<
     }> & {
       paths: Array<TopicSubPath>;
     },
-    _cancel,
-    _meta,
+    _cancel: (reason?: any) => void,
+    _meta: Record<string, JSONValue>,
     ctx: ContextTimed,
   ): AsyncGenerator<ClientRPCResponseResult<AuditEventSerialized>> {
     const { audit }: { audit: Audit } = this.container;
```

Comment:
I'm a bit confused here since I didn't work directly on js-rpc.

But if a handler receives a `cancel` function. What exactly is it cancelling?

Is it cancelling itself? I find this API a bit strange.

From first principles, the idea is that a caller can "abort" its call, by using the controller to abort: `controller.abort(reason);`

This is transmitted to the handler side, and tells the handler that it is aborted. You can check synchronously with `ctx.signal.aborted`, and then the handler essentially deals with this problem gracefully.

In a single-process context, if the caller proceeds to call `cancel`, the caller's promise is rejected with the cancellation reason immediately, but the handler side may continue to function (this example comes from `cancellable.test.ts` in `js-contexts`:

```ts
      const fCancellable = cancellable(f);
      // Signal is aborted afterwards
      const pC1 = fCancellable();
      pC1.cancel('cancel reason');
      await expect(pC1).rejects.toBe('cancel reason');
```

Therefore something similar happens across multi-process context as in the case of RPC. Essentially the call site rejects immediately with a reason, but the handler is able to at their discretion to deal with how it wants to cancel.

The problem right now I see in PK's rpc is that it does not seem that this abort signal is being properly handled in every case. The patterns used to handle abort signal from the caller side should be studied in `js-contexts` and `js-rpc` tests should be representing this correctly.

Now for the `cancel` function on the handler side, I don't quite understand this... why does this exist?

If it is to "cancel" the handling, I would argue that you should just raise an exception, or return early, there's no need to have a `cancel` method. What exactly is it cancelling?

I think this concept generally makes sense under unary handler and server streaming handlers.

In the case of client streaming and duplex streaming, it can be get a bit more complicated. Because now the client is also sending a stream, it's not just the case where the server side is handling a request and returning 1 response or multiple responses, but the client side can send multiple requests, and in cases like duplex streaming, it can also be interleaved order of request, response, response, request request... etc, like a "conversation". That's what I'm referring to as a RPC conversation.

I remember an older issue in js-rpc where I pointed out that the implementation of js-rpc is missing client side transmission of errors: https://github.com/MatrixAI/js-rpc/issues/18.

This could mean that one could also cancel from the server side handling. That would indicate the idea that the SERVER no longer cares about the client's next request. This is important because normally under unary and server streaming, cancelling means "as a client I do not care about your next response", whereas in the case of client streaming and duplex streaming, cancelling from the server side means "as the server I do not care about your next request". Swap request/response for just "message" and you get the idea.

However if you cancel, then you want to make sure that you're also getting the abort reason properly. This is not currently documented or demonstrated in the js-rpc README.

@aryanjassal I expect you to update the README to explain the new implementation and model checked tests in js-rpc to achieve this.

Reply:
This needs to be a new issue in js-rpc to refactor the `cancel` parameter being passed into all the high level handlers. Remove from unary and server streaming, while changing the implementation for client streaming and duplex streaming. You have to change it so that client streaming and duplex streaming on the call-site also gives back a `ContextCancellable`.

Reply:
This needs to be solved with: https://github.com/MatrixAI/js-rpc/issues/18

Reply:
I think it's not clear in the js-rpc documentation how you're supposed to properly "cancel" the server handler for all 4 call types. Even in the case of unary calls, you don't get back a `PromiseCancellable` where you can do:

```ts
const callP = rpcClient.methods.testMethod();
callP.cancel();
```

Reply:
@aryanjassal this needs a new issue.

1. We want to use `PromiseCancellable` on the client callers.
2. This means it should be possible to do `callP.cancel(reason);`
3. This should translate into a stream cancel reason using the `reasonToCode`.
4. On the server side where `codeToReason` occurs, that should try to recover an exception object or a generic exception object. (This will require indexing all exceptions by code numbers).
5. The handler on the server side should then see it as part of the `ctx.signal.reason`.
6. High level handlers on the server side should not have access to the `cancel` parameter.
7. @tegefaulkes mentioned that the `cancel` parameter is being used by isogit. However this shouldn't be done this, way I believe it should be possible to bridge the `ctx.signal` into the cancellation of the isogit http stream: https://github.com/matrixai/polykey/pull/846#issuecomment-2544047314, it shouldn't be using the `cancel` parameter directly.
8. I don't know if the `cancel` parameter is needed at the handler level at all.

Reply:
On point 7, it's not being used by isogit. It's just Isogit has no built in cancellation so we'd need to use the `cancel()` to propagate the cancellation to the stream layer to end the cloning and pulling operation earlier.

Reply:
https://github.com/MatrixAI/js-rpc/issues/77